enhancements/machine-api/control-plane-machine-set.md

```diff
@@ -0,0 +1,801 @@
+In particular it will also allow them to have different specs across their control plane.
+While in the normal case this is discouraged, this could be used for short periods to test new configuration before
+rolling the updated configuration to all Control Plane Machines.
+
+Note that the OnDelete strategy is similar to the RollingUpdate strategy in that it needs to be able to add new
+Machines to the cluster to function. It can be thought of as a slower, more controlled RollingUpdate strategy where the
+user decides when each Control Plane Machine is replaced, and marks it for replacement by deleting it.
+Otherwise the replacement is identical to the RollingUpdate strategy.
+
+#### Protecting etcd quorum during update operations
+
+With the introduction of the [etcd protection enhancement](https://github.com/openshift/enhancements/pull/943), the
+ControlPlaneMachineSet does not need to observe anything etcd related during scaling operations.
+In particular, it is the etcd operators responsibility to add Machine deletion hooks to prevent Control Plane Machines
+from being removed until they are no longer needed.
+When the `ControlPlaneMachineSet` controller observes that a newly created Machine is ready, it will delete the old
+Control Plane Machine signalling to the etcd operator to switch the membership of the etcd cluster between the old and
+new instances.
+Once the membership has been switched, the Machine deletion hook will be removed on the old Machine, allowing it to be
+removed by the Machine controller in the normal way.
+
+#### Removing/disabling the ControlPlaneMachineSet
+
+As some users may want to remove or disable the ControlPlaneMachineSet, a finalizer will be placed on the
+`ControlPlaneMachineSet` to allow the controller to ensure a safe removal of the ControlPlaneMachineSet, while leaving
+the Control Plane Machines in place.
+
+Notably it will need to ensure that there are no owner references on the Machines pointing to the ControlPlaneMachineSet
+instances. This will prevent the garbage collector from removing the Control Plane Machines when the
+`ControlPlaneMachineSet` is deleted.
+
+If users later wish to reinstall the `ControlPlaneMachineSet`, they are free to do so.
+
+#### Installing a ControlPlaneMachineSet within an existing cluster
+
+When adding a ControlPlaneMachineSet to a new cluster, the end user will need to define the `ControlPlaneMachineSet`
+resource by copying the existing Control Plane Machine ProviderSpecs.
+Once this is copied, they should remove the failure domain and add the desired failure domains to the `FailureDomains`
+field within the ControlPlaneMachineSet spec.
+
+To ensure adding a `ControlPlaneMachineSet` to the cluster is safe, we will need to ensure via a validating webhook
+that the replica count defined in the spec is consistent with the actual size of the control plane within the cluster.
```

Comment:
is it too difficult to let them create an empty controlplanemachineset and then populate it properly for them?

Reply:
This was previously discussed (https://github.com/openshift/enhancements/pull/1008#discussion_r790866005) but I'm not sure it made it into the enhancement. My main concern here is that there are many scenarios where the config may have changed or may have been broken in some way.

For example, if a customer resized their control plane out of band to add more capacity, then forgot to update the Machine specs to represent the actual state. Now if we infer from the Machines what their configuration is, we may then resize their control plane back down to the previous size, possibly causing a cluster failure as the smaller instances cannot cope with the load on them. That's then us breaking their cluster through automation. If they were the one to come up with the configuration, that's on them for breaking the configuration.

If we were to go down the route of auto populating this, I think we would need some sort of check box that the customer acknowledges the inferred configuration and says yep, go ahead, make that the actual state in case it's not, I accept the responsibility. Any idea how we might do that? I can add it as a future improvement to the project (I don't want to extend the scope right now, I think it's already stretching what we can achieve in 4.11)

Reply:
if ControlPlaneMachineSet had an active/inactive field (which i think i saw some need for elsewhere in the EP while i was reading it) you could create the ControlPlaneMachineSet in an Inactive state, and the admin can review what it is going to do before marking it active/enforced.

Reply:
(and yes, keeping this out of scope for 4.11 sounds perfectly reasonable)

Reply:
I am very excited to see this enhancement and look forward to testing it out.  I assume this means the limitation of 3 control plane nodes is being removed as well but did not see it explicitly listed.

I second the idea of an active/inactive flag to satisfy the potential issue of out of band changes to the control plane.  Put that reconciliation back in the admin.

Reply:
> if ControlPlaneMachineSet had an active/inactive field

This was mentioned in a previous conversation though I didn't think it had made it in. We have been using `paused` in the upstream Cluster API community so would likely copy that pattern, but it's effectively the same

> I assume this means the limitation of 3 control plane nodes is being removed as well but did not see it explicitly listed.

This is listed as a non-goal, we will only support vertical scaling at first and not horizontal scaling

> I second the idea of an active/inactive flag to satisfy the potential issue of out of band changes to the control plane. Put that reconciliation back in the admin.

You can also effectively do this by using the OnDelete strategy, which won't take action until a user deletes one of the existing Control Plane Machines